### PR TITLE
[WIP] [Kernel]Make kernel lib suit the standalone.

### PR DESCRIPTION
**Pull request by horizonzy**

### CODE REVIEW

File: connectors/flink/src/test/java/io/delta/flink/source/internal/KernelMetadataUtils.java [hunk 1]
```diff
@@ -92,24 +93,7 @@ public String getString(int rowId) {
                 }
             },
             Optional.ofNullable(1234L),
-            new MapValue() { // conf
```
Comment (vkorukanti) [inline connectors/flink/src/test/java/io/delta/flink/source/internal/KernelMetadataUtils.java:95]: make the code cleanup here as a separate PR.

File: kernel/kernel-api/src/main/java/io/delta/kernel/engine/FileSystemClient.java [hunk 2]
```diff
@@ -77,4 +77,6 @@ CloseableIterator<ByteArrayInputStream> readFiles(
      * @throws IOException for any IO error.
      */
     boolean mkdirs(String path) throws IOException;
+
+    boolean delete(String path) throws IOException;
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/engine/FileSystemClient.java:81]: This interface method and its default implementation can be a separate PR.

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ArrayColumnReader.java [hunk 3]
```diff
@@ -79,14 +89,16 @@ private static Converter createElementConverter(
 
         checkArgument(typeFromFile.getFieldCount() == 1,
                 "Expected exactly one field in the array type, but got: " + typeFromFile);
-        GroupType repeatedGroup = typeFromFile.getType(0).asGroupType();
-
-        // TODO: handle the legacy 2-level list physical format
-        checkArgument(repeatedGroup.getFieldCount() == 1,
+        Type type = typeFromFile.getType(0);
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ArrayColumnReader.java:92]: Is this supporting a legacy list formats? If yes, I think there are more cases to handle. Also checkout the `ParquetFielReaderSuite.scala` (`read parquet file generated by parquet-thrift`) for disabled tests which can be enabled once this is fixed.

Also this can be a separate PR.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 4]
```diff
@@ -60,6 +60,31 @@ public class TableConfig<T> {
             "needs to be a positive integer."
     );
 
+    public static final TableConfig<Long> LOG_RETENTION = new TableConfig<>(
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:63]: These can be added in a separate PR.

**Review by vkorukanti — CHANGES_REQUESTED** [review]
Hi @horizonzy, Thanks for this PR. Overall it has lot of pieces in it. I commented on the few code pointers which can be extracted out into separate PRs.

**Comment by horizonzy** [comment]
Split PRs:
#3209 
#3210
#3212 
#3213

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ArrayColumnReader.java (hunk 3 above)
Comment (horizonzy) [reply kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ArrayColumnReader.java:92]: Yes, here I just make a patch to work around the issue. We can open a new issue to trace the problem, I will attach the test table, it's easy to reproduce the issue.
